### PR TITLE
Select complete multi-line statement when running run-in-console action without selection

### PR DESCRIPTION
**Pull request by BoPeng**
https://github.com/jupyterlab/jupyterlab/pull/4330#issuecomment-500031828

## Code changes

See updated algorithm at the end of [this comment](https://github.com/jupyterlab/jupyterlab/pull/6515#issuecomment-501380297).

## User-facing changes

Using `run-in-console` action to run through the cell

See updated screenshots [here](https://github.com/jupyterlab/jupyterlab/pull/6515#issuecomment-501400917)

## Potential risk of executing unintended code

Jason [pointed out](https://github.com/jupyterlab/jupyterlab/pull/6515#issuecomment-500610279) two potential problems with this PR. Namely there is no indication of what would be executed before the action, and the possibility of executing large trunk of unintended code. The PR has since been updated to closely mimic RStudio's behavior in case of syntax errors, which will **only include code before the current line if complete statements are identified from the beginning of the cell and the current line falls into a complete multi-line statement**. This has worked pretty well for all cases I have tested. 

## Potential performance problem:

Each time the algorithm will start from the top of cell, send code to the kernel to find a complete statement that contains the current line. This can be a bit slow for large cells but Jupyter does not encourage such cells anyway. 

@jasongrout please have a look.

### CODE REVIEW

File: packages/notebook-extension/src/index.ts [hunk 1]
```diff
@@ -974,14 +974,59 @@ function addCommands(
...
-        // no selection, submit whole line and advance
-        code = editor.getLine(selection.start.line);
+        // no selection, find the complete statement around the current line
         const cursor = editor.getCursorPosition();
-        if (cursor.line + 1 !== editor.lineCount) {
-          editor.setCursorPosition({
-            line: cursor.line + 1,
-            column: cursor.column
-          });
+        let srcLines = editor.model.value.text.split('\n');
```
Comment (BoPeng) [inline packages/notebook-extension/src/index.ts]: @jasongrout Should I also consider `\r\n` stuff here?

Comment (jasongrout) [reply packages/notebook-extension/src/index.ts]: I think the editor automatically converts \r\n to \n, so you can consistently use \n. @ian-r-rose?

**Comment by jasongrout** [comment]
This seems very convenient. My big concern in the idea here is that this might run out of control. For example, if I accidentally try to execute a non-code line (perhaps just a line of text in a text file that also happens to contain python code), it may suck up the whole file and try to execute it? Or if I have a missing closing paren, will it mistakenly try to execute the entire file from the cursor to the end?

**Comment by jasongrout** [comment]
It would be great if there were some ui indication of what code was going to be executed before execution. I'm not sure how to do that in an efficient way, though.

File: packages/notebook-extension/src/index.ts (hunk 1 above)
Comment (ian-r-rose) [reply packages/notebook-extension/src/index.ts]: In the notebook we never use `\r\n`, so I don't think it should ever be in there unless you use some alternative notebook editing software.

We make an effort to keep convert `\r\n` to `\n` when editing text files, and then restore them upon saving.

**Comment by BoPeng** [comment]
I had the same worry. Basically, if the current line is invalid, it is unlikely that adding lines before or after it will make it run. If the current line is part of a huge multi-line function, then yes the entire function will be executed.

It would be great if, for example, when I click a line, there is some indication of the boundary of a complete statement. This is however a nice feature by itself and can be added later.

**Comment by BoPeng** [comment]
https://github.com/jupyterlab/jupyterlab/pull/6515/commits/7be4a2520f9d3c18033c942f19e94ac54d02cd57 handles empty lines more cleverly. More specifically,

1. If `run-in-console` starts from an empty line, it moves down to find the first non-empty line.
2. If there are empty lines following end of the statement, it moves to the beginning of the non-empty line.

These are consistent with RStudio behaviors.

**Comment by BoPeng** [comment]
@jasongrout It turns out that the problem I had with IPython kernel was due to a bug that has been fixed in the latest version of `ipython` (7.5.0). I have submitted a [patch](b9d3164) to remove my walk-around. This PR works pretty well for all the kernels I have tested (I have updated the ticket with a screenshot for R) and should make `run-in-console` much easier to use so I highly recommend its inclusion in 1.0.

**Comment by jasongrout** [comment]
An option is also to put this behind an option. In general, I tend to favor predictability and lower chance of failure over trying to be smart by default, especially when it comes to guessing user intentions with little chance to recover from small user errors. However, providing an option where the user can opt into something that tries to be smarter great.

**Comment by afshin** [comment]
We discussed this in the weekly call and proposed the following behavior:

* There could be setting that indicates the number of lines you want to read ahead with this behavior.
* The default number could be something small, like `10` lines.
* If a user wants to disable this, it can be set to `0`.
* If a user has a higher risk threshold, the number could be very large.

What do you think?

**Comment by jasongrout** [comment]
I think that is a nice balance of intelligence vs complexity and the possibility of user error derailing things.

**Comment by BoPeng** [comment]
Stepping through code naturally means "stepping by full statements" in any syntax-aware IDE. RStudio does this nicely, so do PyCharm and other Python IDEs. The `run-in-console` feature has been introduced to `sos notebook` for a while and the most complaint I have heard was that it did not work as well as RStudio when stepping through code, and I am glad that we are making JLab on par with RStudio in this aspect.

I think the biggest concern of this PR is that it could probe forward and backward too much and execute unintended codes. In an extreme case, one could step through the code and get some code included/executed again when executing a new line. To assess the risk of this I have played with Rstudio more and have some interesting findings. 

When the current line is complete by itself even if it is part of a larger complete statement, e.g. 
```
a = c(1,
 2   <- cursor here.
, 3)
```
Currently this PR will execute `2` by itself, while RStudio will execute the entire statement in a similar case, regardless of number of lines for the statement.

Very interestingly, the behavior of RStudio will change with/without an invalid statement before the current statement. I I mean, `Ctrl-Enter` at the `6` line will send only `6` in the case of

```
a <- c(
 
c = c(
  6 
)
```
and the complete 
```
c = c(
  6 
)
```
in the case of 
```
# a <- c(
 
c = c(
  6 
)
```

Moreover, in the case of

```
a <- c(
  
c = c(  <- cursor here
  6  ) 
```
Rstudio will be able to identify `c = c( 6)`, but fail to get the complete statement  in the case of
```
a <- c(
  
c = c(  
  6  ) <- cursor here
```
while this PR can.

In a similar case, our PR could get

```
'''
wrong  = [ 
b = '''
```
when evaluating a cell with code

```
a = '''
my string
'''
wrong  = [     <- cursor here
b = '''
my another string 
'''
```
while RStudio would handle it correctly.

My conclusion is that the algorithm for this PR is not ideal and the cited problems could be resolved by a simpler yet safer algorithm that **are more careful in search backward**

1. Starting from the current line, search **forward** for complete statement, submit if found.
2. Starting from the first line and try to find complete statements, if a complete statement that contains the current line is found, submit the statement.
3. Otherwise send the current line.

**Comment by BoPeng** [comment]
The new algorithm has been implemented, which I believe should be much safer, at least as safe as RStudio.

image

**Comment by jasongrout** [comment]
A little simplified await construct:

```patch
diff --git a/packages/notebook-extension/src/index.ts b/packages/notebook-extension/src/index.ts
index ea848dc91..1cb7c39ad 100644
--- a/packages/notebook-extension/src/index.ts
+++ b/packages/notebook-extension/src/index.ts
@@ -994,14 +994,9 @@ function addCommands(
             code: code
           };
 
-          let completed = false;
-          await current.context.session.kernel
-            .requestIsComplete(msg)
-            .then(reply => {
-              completed = reply.content.status === 'complete';
-            });
+          let reply = await context.session.kernel.requestIsComplete(msg);
 
-          if (completed) {
+          if (reply.content.status === 'complete') {
             if (curLine < lastLine) {
               // we find a block of complete statement containing the current line, great!
               while (
```

**Comment by jasongrout** [comment]
I notice that you have the heuristic of stopping when you hit a whitespace line. Is that encoding too much of a syntax assumption about the language?

**Comment by BoPeng** [comment]
The empty lines (empty or with only space and tabs) are handled as follows:

1. If the starting line is empty, the algorithm moves to the first non-empty line. 
2. It includes empty lines when searching for complete statement.
3. If a complete statement is found and executed, the algorithm moves the cursor to the non-empty line after the statement.

The behavior saves the trouble of sending empty lines to the console (which should be meaningless in any language) and makes it easier to step through code with empty lines. As for syntax assumption, 2 and 3 should be safe, and the 1 could be problematic with valid empty line such as
```
a = [1,
  <- cursor here.
2]
```
However, because "valid empty line" is unlikely to be the starting line, JLab will correctly identify the statement and submit
```
a = [1

2]
```
after the cursor is moved to the last line. It would send 
```
2]
```
when statements before the assignment is invalid, but it is still better than sending an empty line to the console.

So I think the behavior should be safe for any language.

**Comment by jasongrout** [comment]
Thanks. I think this is ready to go in now, and let's try it out. If it turns out to be too smart (i.e., if it is not giving people what is wanted), we can introduce an option to turn it off.

**Comment by BoPeng** [comment]
Great. Thanks!

**Comment by jasongrout** [comment]
Yes, this PR is in jlab 1.0.4. I just checked and it works for me: putting my cursor on the line `a = [1,` runs both lines: `a = [1, 2]`.

As an aside, Shift-G for the shortcut? As in you can't type capital G in the code cell? Is this perhaps related? (I use Ctrl G so I can still type a capital G)

**Comment by beaulucas** [comment]
@jasongrout 

Thanks for fast response, I diagnosed the issue on my end. I was launching JupyterLab from a 2.7 environment, and had the impression that user settings were kernel-dependent (1.0.4 is on my 3.7.2 environment). Still getting a hang of the ropes on this. Coming from RStudio. 

You make a great point on the Shift remark... Reverting back to CTRL.

**Comment by jasongrout** [comment]
Great to hear it's working.

> Still getting a hang of the ropes on this. Coming from RStudio.

Welcome!

**Pull request merged**